src/libsync/vfs/cfapi/cfapiwrapper.cpp

```diff
@@ -124,21 +124,87 @@ void CALLBACK cfApiFetchDataCallback(const CF_CALLBACK_INFO *callbackInfo, const
+
+        // On slow connection we might not be able to fill one block complete.
+        // Threrefore we add the data to protrudingData and wait for the next run. If this
+        // was the last run, the protrudingData will get flushed after the hydration job emitted it's
+        // finished signal
+        if (data.size() + protrudingData.size() < cfapiBlockSize) {
+            protrudingData.append(data);
+            return;
+        }
+
+        int dataIndex = 0;
+        QByteArray dataToSend;
+        // Is there any data left from last run?
+        if (!protrudingData.isEmpty()) {
+            dataToSend.append(protrudingData);
+            protrudingData.clear();
+
+            // Fill up remaining first block if possible
+            const auto fillupSize = cfapiBlockSize - dataToSend.size();
+            const auto missingSize = std::min(fillupSize, data.size());
+            if (missingSize > 0) {
+                dataToSend.append(data.left(missingSize));
+                dataIndex += missingSize;
+            }
+        }
+
+        // Do we still have data to send?
+        if (dataIndex < data.size()) {
+            // Extract blocks from data if possible
+            const auto blocksCount = (data.size() - (dataIndex + 1)) / cfapiBlockSize;
+            if (blocksCount != 0) {
+                dataToSend.append(data.mid(dataIndex, cfapiBlockSize * blocksCount));
+                dataIndex += cfapiBlockSize * blocksCount;
+            }
+
+            // Do we still have any protruding data left? If yes, save this data for the next run and send it then
+            if (dataIndex < data.size()) {
+                protrudingData = data.right(data.size() - dataIndex);
+                qDebug(lcCfApiWrapper) << "Save protruding data for next run. Size:" << protrudingData.size();
+            }
+        }
+
+        // Now if we have any data accumulated, send it
+        if (dataToSend.size() > 0) {
+            // Did we arrive the end of the data? If yes, also append the protruding data
+            if (!protrudingData.isEmpty() && callbackInfo->FileSize.QuadPart == offset + dataToSend.size() + protrudingData.size()) {
+                dataToSend.append(protrudingData);
+            }
+            qDebug(lcCfApiWrapper) << "Send data block. Size:" << dataToSend.size();
+            sendTransferInfo(dataToSend, offset);
+            offset += dataToSend.size();
+        }
     });
 
     QObject::connect(vfs, &OCC::VfsCfApi::hydrationRequestFinished, &loop, [&](const QString &id, int s) {
+        qDebug(lcCfApiWrapper) << "Hydration finished for request" << id;
         if (requestId == id) {
+//            // Exit the loop before we flush the protruding data. It might happen that there is still some data
+//            // that needs to processed before we send the protruding data. This call will execute the pending events.
+//            loop.quit();
+
+//            if (!protrudingData.isEmpty()) {
```

Comment:
I'd expect this to still be necessary, isn't it? Although I'd expect to do it after the `loop.exec()` call. Seems like a more natural place to me, to make sure you catch all missing reads first.

Reply:
What is still necessary? The `loop.quit()`? It is still there. It's just there were it originally was. The flushing of the protruding data is now done in `readyRead()`.

Reply:
No I meant the flushing. But I guess that explains why this code is fairly complicated at that point. It's too much in the readyRead in my opinion.

Reply:
The problem with this `OCC::VfsCfApi::hydrationRequestFinished()` method is that it might get called before the socket has finished reading all the bytes. That's why we moved the flushing of the protruding data into the `readyRead()` function. I'm not sure why this happens. I was thinking because `loop.quit()` needs to be executed before the flushing of the protruding data to execute pending events. But I'm not sure about it. Had a discussion about that with @allexzander in the chat.

Reply:
As the code currently is, it's not necessary to flush anything in `::VfsCfApi::hydrationRequestFinished`.